compiler/qsc_rca/src/common.rs

```diff
@@ -0,0 +1,354 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use crate::QuantumProperties;
+use indenter::Indented;
+use qsc_data_structures::index_map::IndexMap;
+use qsc_fir::{
+    fir::{
+        CallableDecl, ExprId, ExprKind, Functor, ItemId, NodeId, PackageId, PackageLookup, Pat,
+        PatId, PatKind, Res, StoreItemId, UnOp,
+    },
+    ty::Ty,
+};
+use rustc_hash::FxHashMap;
+use std::fmt::{Debug, Formatter};
+
+use crate::{ComputeKind, ValueKind};
+
+/// Aggregates two compute kind structures returning the resulting compute kind.
+#[must_use]
+pub fn aggregate_compute_kind(basis: ComputeKind, delta: &ComputeKind) -> ComputeKind {
+    let ComputeKind::Quantum(delta_quantum_properties) = delta else {
+        // A classical compute kind has nothing to aggregate so just return the base with no changes.
+        return basis;
+    };
+
+    // Determine the aggregated runtime features.
+    let runtime_features = match basis {
+        ComputeKind::Classical => delta_quantum_properties.runtime_features,
+        ComputeKind::Quantum(ref basis_quantum_properties) => {
+            basis_quantum_properties.runtime_features | delta_quantum_properties.runtime_features
+        }
+    };
+
+    // Determine the aggregated value kind.
+    let value_kind = match basis {
+        ComputeKind::Classical => delta_quantum_properties.value_kind,
+        ComputeKind::Quantum(basis_quantum_properties) => aggregate_value_kind(
+            basis_quantum_properties.value_kind,
+            &delta_quantum_properties.value_kind,
+        ),
+    };
+
+    // Return the aggregated compute kind.
+    ComputeKind::Quantum(QuantumProperties {
+        runtime_features,
+        value_kind,
+    })
+}
+
+#[must_use]
+pub fn aggregate_value_kind(basis: ValueKind, delta: &ValueKind) -> ValueKind {
+    match delta {
+        ValueKind::Static => basis,
+        ValueKind::Dynamic => ValueKind::Dynamic,
+    }
+}
+
+/// The index corresponding to an input parameter node.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct InputParamIndex(usize);
+
+/// An input parameter node.
+#[derive(Clone, Debug)]
+pub struct InputParam {
+    pub index: InputParamIndex,
+    pub pat: PatId,
+    pub ty: Ty,
+    pub node: Option<NodeId>,
+}
+
+pub fn derive_callable_input_params(
+    callable: &CallableDecl,
+    pats: &IndexMap<PatId, Pat>,
+) -> Vec<InputParam> {
+    let input_elements = derive_callable_input_pattern_elements(callable, pats);
+    let mut input_params = Vec::new();
```

Comment:
The length of `input_params` is known based on the number of `input_elements`. Initialize it with capacity: `let mut input_params = Vec::with_capacity(input_elements.len());`